Stream concurrently

nomisRev opened the pull request:
- Add `concurrently`
- Add test case to cover "propagate error from pure interrupted stream", and fixes bug

nomisRev commented on arrow-fx-coroutines/src/main/kotlin/arrow/fx/coroutines/Bracket.kt:
```diff
@@ -8,8 +8,12 @@ import kotlin.coroutines.intrinsics.startCoroutineUninterceptedOrReturn
+  object Cancelled : ExitCase() {
+    override fun toString(): String = "ExitCase.Cancelled"
+  }
   data class Failure(val failure: Throwable) : ExitCase()
```
The reason for this override was to remove the `@hashCode` postfix from `Completed` & `Cancelled` objcets which were polluting the logs a bit.

JorgeCastilloPrz replied:
I know

nomisRev commented on arrow-fx-coroutines/src/main/kotlin/arrow/fx/coroutines/SuspendConnection.kt:
```diff
@@ -17,6 +17,8 @@ inline class CancelToken(val cancel: suspend () -> Unit) {
 
   suspend fun invoke(): Unit = cancel.invoke()
 
+  override fun toString(): String = "CancelToken(..)"
```
No, we cannot show anything useful about the lambda.
I overwrote this also to not pollute the logs with useless info.

`CancelToken(cancel = suspend () -> Unit)` becomes `CancelToken(..)`.

nomisRev commented on arrow-fx-coroutines/src/main/kotlin/arrow/fx/coroutines/stream/Chunk.kt:
```diff
@@ -143,7 +143,7 @@ abstract class Chunk<out O> {
     when {
       n <= 0 -> empty()
-      n == 1 -> fromNullable(firstOrNull())
+      n == 1 -> just(get(0))
```
No, it's not unsafe since we check the size first.

JorgeCastilloPrz replied:
Ah 🤦

aballano replied:
The `n <= 0` will cover against wrong params, but if there's an empty Chunk and we request 1, won't this crash here?

nomisRev commented on arrow-fx-coroutines/src/main/kotlin/arrow/fx/coroutines/stream/Chunk.kt:
```diff
@@ -154,7 +154,7 @@ abstract class Chunk<out O> {
     when {
       n <= 0 -> empty()
-      n == 1 -> fromNullable(lastOrNull())
+      n == 1 -> just(get(size() - 1))
```
Covered by `n <= 0 -> empty()`

nomisRev commented on arrow-fx-coroutines/src/main/kotlin/arrow/fx/coroutines/stream/Compiler.kt:
```diff
@@ -106,7 +106,7 @@ internal fun <O> interruptBoundary(
         is Pull.Eval.CloseScope -> Pull.Eval.CloseScope(
           close.scopeId,
-          Pair(interruptedScope, close.interruptedScope?.second),
+          Pair(interruptedScope, interruptedError),
```
We were passing along the wrong exception here.
This fixes https://github.com/arrow-kt/arrow-fx/pull/210/files#diff-ce71ba0355e5832a46e62f6d1ee87360R249.

The problem there was that we need to pass along the `interruptedError`, to re-raise the error later on.


If we look at how `flatMap` is implemented, we can see that it does `unconsOrNull`, which pulls the next `Chunk<O>.

Then it'll transform the `tail` with this method `interruptBoundary`:
 - in the case the `tail` was not finished (`Pull.Result`), it will either `Intterupt` or `Fail`
 - When we encounter `CloseScope`, we're going to set our current  `Interuption Context` to our `CloseScope`.
 - When we encounter a case that produces a new value i.e. `Effect` we just evaluate it and move to the next step.

JorgeCastilloPrz replied:
Right, I think I should have been more clear with my question. Isn't there a chance for interruptedError to be `null` here for the `CloseScope`? what'd be the case when that happens?

nomisRev replied:
Ah, yes there is a chance for it to be `null`.
If it's `null` then we consider it a graceful interruption.

`Stream.sleep(1.second).interruptWhen { Right(sleep(500.milliseconds)) }`
 vs

`Stream.sleep(1.second).interruptWhen { sleep(500.milliseconds); Left(RuntimeException("Boom!")) }`

JorgeCastilloPrz replied:
Perfect, thanks for clarifying!

nomisRev commented on arrow-fx-coroutines/src/main/kotlin/arrow/fx/coroutines/stream/Stream.kt:
```diff
@@ -1322,6 +1323,69 @@ import kotlin.random.Random
+   * with an error.
+   *
+   * Upon finalization, the resulting stream will interrupt the background stream and wait for it to be
+   * finalized.
```
No, there is no difference from `parJoin`.

```
val one = Stream(1)
val other = Stream.bracket({ Unit }, { sleep(1.second) }).flatMap { never<Unit>() }

one.concurrently(other)
```

`one` finishes immediately, but it waits until `other` is finished finalizing.
So here it will wait until the `sleep(1.seconds)` is finished, before continuing to for example an `.append { }`.

JorgeCastilloPrz replied:
👍 so what'd be the rationale for offering both?

nomisRev replied:
In `parJoin` all inner streams run as child processes of the outer parent process.
With `concurrent` `this` is parent stream.

JorgeCastilloPrz replied:
Ok, then there is a relevant difference 👍

nomisRev commented on arrow-fx-coroutines/src/test/kotlin/arrow/fx/coroutines/stream/Arb.kt:
```diff
@@ -43,16 +42,21 @@ inline fun <reified A> Arb.Companion.array(
-fun <A, B> arrayChunkGenerator(
-  gen: Gen<A>,
+@PublishedApi
+internal fun <A, B> arrayChunkGenerator(
```
Right, I can move these to a separate PR. No problem. Should we figure out some guidelines?

JorgeCastilloPrz replied:
I'm not sure what @aballano is used to exactly, but on my side I'm used to [Gitflow](https://www.atlassian.com/git/tutorials/comparing-workflows/gitflow-workflow) or similar variants of it. However, that's something to be agreed with everybody, I'd not block this PR for a decision like that. These flows also cover things like release branches that we could use for our docs thing discussed yesterday.

That said, the single responsibility for PRs applies the same way for any workflows, since it's more like a matter of making those as atomic as possible, simply to make it easier for reviewers. Imho, ideally every PR would cover a single topic, and be as atomic as you can for segregation. (Almost any PR can be chunked on very differentiated steps of a whole).

That said also, If a "cowboy" change is quite related to the changes on the diff, then I guess it's fine, but there is always a point where you start feeling like it's too much to include. At that point it's probably better as a separate PR with its proper description / discussion. I know common sense is not a very common sense, but this would be more on every one to consider and evaluate.

rachelcarmena replied:
> I know common sense is not a very common sense, but this would be more on every one to consider and evaluate.

"common sense" :clap: and I'd add "empathy": _could someone review this PR easily?_

aballano replied:
IMO changes like this in tests for small PRs like this one shouldn't matter (at least to me personally) as long as this doesn't turn into the main changes of the PR

nomisRev commented on arrow-fx-coroutines/src/test/kotlin/arrow/fx/coroutines/stream/ConcurrentlyTest.kt:
```diff
@@ -0,0 +1,120 @@
+import io.kotest.property.Arb
+import io.kotest.property.arbitrary.int
+
+class ConcurrentlyTest : StreamSpec(spec = {
```
Correct, that test is indeed missing!

nomisRev replied:
@JorgeCastilloPrz there are some other cancellation tests still missing.

Shall we add them all in a separate PR, or shall we include them here?

JorgeCastilloPrz replied:
If they are not cancellation for `concurrently` I'd say different PR, so we can get this one in without depending on that.

aballano commented on arrow-fx-coroutines/src/test/kotlin/arrow/fx/coroutines/predef-test.kt:
```diff
@@ -203,14 +203,16 @@ internal fun <A> Either<Throwable, A>.suspended(): suspend () -> A =
-  try {
-    val a = executable.invoke()
-    fail("Expected an exception but found: $a")
+inline fun <A> assertThrowable(executable: () -> A): Throwable {
```
I discovered that we can call `shouldThrow` instead, do we need this function at all?

nomisRev replied:
If it offers the same behavior then we definitely don't need this one!
I replaced this instead of an `inline reified` one since I'm always just capturing `Throwable` and then use `shouldBe` instead.

So I use it the same as `Either.catch { } shouldBe Either.Left(e)` vs `assertThrowable { } shouldBe e`

aballano commented on arrow-fx-coroutines/src/main/kotlin/arrow/fx/coroutines/stream/Chunk.kt:
```diff
@@ -143,7 +143,7 @@ abstract class Chunk<out O> {
     when {
       n <= 0 -> empty()
-      n == 1 -> fromNullable(firstOrNull())
+      n == 1 -> just(get(0))
```
one comment lost in the discussion:

The n <= 0 will cover against wrong params, but if there's an empty Chunk and we request 1, won't this crash here?

nomisRev replied:
Yes... that means our `Arb` is not complete. I think we need to add `Chunk.empty` as an edgecases.
I'll move this to a separate PR.

nomisRev commented:
I moved the changes to `Chunk` to a different PR already merged #215.
& I moved the fix related to `interruption` & `bracket` from this PR to #222.

When adding the suggested test related to cancellation, I found it failing.
There seems to be a flaky race condition somewhere when launching and immediately canceling.
I see the error occur every ~ 100-200 iterations.
We can either block this PR until that is resolved or create a separate ticket for it. Thoughts?

cc\\ @JorgeCastilloPrz @aballano

aballano commented:
> I moved the changes to `Chunk` to a different PR already merged #215.
> & I moved the fix related to `interruption` & `bracket` from this PR to #222.
> 
> When adding the suggested test related to cancellation, I found it failing.
> There seems to be a flaky race condition somewhere when launching and immediately canceling.
> I see the error occur every ~ 100-200 iterations.
> We can either block this PR until that is resolved or create a separate ticket for it. Thoughts?
> 
> cc\ @JorgeCastilloPrz @aballano

IMO we can create an issue apart, no need to block all this work for a flaky test that we can fix later

JorgeCastilloPrz commented:
Sounds good 👍

nomisRev commented:
@aballano I moved `assertCancellable` utilities to `CancellationTest` in the last commit in this PR.

nomisRev commented:
@aballano
